src/index/index.c

```diff
@@ -0,0 +1,141 @@
+#include "index.h"
+
+int compareNodes(const void *a, const void *b) {
+  return ((Node*)a)->id - ((Node*)b)->id;
+}
+
+int compareStrings(void *a, void *b, void *ctx) {
+  return strcmp(((SIValue*)a)->stringval, ((SIValue*)b)->stringval);
+}
+
+// TODO this seems inefficient
+int compareNumerics(void *p1, void *p2, void *ctx) {
+  double a, b;
+  SIValue_ToDouble(p1, &a);
+  SIValue_ToDouble(p2, &b);
+  return a - b;
+}
+
+/*
+ * The index must maintain its own copy of the indexed SIValue
+ * so that it becomes outdated but not broken by updates to the property.
+ */
+void cloneKey(void **property) {
+  SIValue *redirect = *property;
+  *redirect = SI_Clone(*redirect);
+}
+
+void freeKey(void *key) {
+  SIValue_Free(key);
+}
+
+// Create and populate index for specified property
+// (This function will create separate string and numeric indices if property has mixed types)
+void indexProperty(RedisModuleCtx *ctx, const char *graphName, AST_IndexNode *indexOp) {
+  const char *index_label = indexOp->target.label;
+  const char *index_prop = indexOp->target.property;
+  LabelStore *store = LabelStore_Get(ctx, STORE_NODE, graphName, index_label);
+
+  LabelStoreIterator it;
+  LabelStore_Scan(store, &it);
+
+  char *nodeId;
+  uint16_t nodeIdLen;
+  Node *node;
+  EntityProperty *prop;
+
+  // We maintain one index per property per type;
+  // these will be created as necessary
+  Index *numeric_index = NULL;
+  Index *string_index = NULL;
+
+  // In the most common case, the node currently being indexed will be placed in the same index as the
+  // previous node, so we will try to optimize for that scenario
+  SIType last_key_type = T_NULL;
+  Index *current_index = NULL;
+
+  int i;
+  int prop_index = 0;
+  while(LabelStoreIterator_Next(&it, &nodeId, &nodeIdLen, (void**)&node)) {
+    // If the sought property is at a different offset than it occupied in the previous node,
+    // then seek and update
+    if (strcmp(index_prop, node->properties[prop_index].name)) {
+      for (i = 0; i < node->prop_count; i ++) {
+        prop = node->properties + i;
+        if (!strcmp(index_prop, prop->name)) {
+          prop_index = i;
+          break;
+        }
+      }
+    }
+    prop = node->properties + prop_index;
+    // This value will be cloned within the skiplistInsert routine if necessary
+    SIValue *key = &prop->value;
+
+    if (key->type ^ last_key_type) {
+      // The current property is of a different type than the last,
+      // or we have not yet constructed an index
+      last_key_type = key->type;
+      if (key->type == T_STRING) {
+        if (string_index == NULL) {
```

Comment:
I don't know if these conditionals will incur a performance penalty from over-eager branch prediction - it might actually be cheaper to build 'string_index' and 'numeric_index' before this loop, then free them if they are of length 0 after the loop.